### ISSUE TITLE
Publish event doesn't work with no error message

### ISSUE THREAD

**Issue by Gianluca-Casagrande-Stiga** [moscajs/aedes#666]
Hi everyone,
I have a mqtts server running 24/7 with about 20 clients which are publishing 500B payload every second.
At a random time (3-5 times a week) the publish event stops to work: from client side there is no error, you can do the publish command. But from the broker side, no publish event is emitted. On the subscriber client obviously you don't receive any message.
There are no aedes error message: the unique solution I've found is to restart the server.
Anyone has the same problem?

I'm using version 0.46.1

thank you

**Comment by robertsLando** [comment]
Could you submit a test code to reproduce the issue?

**Comment by Gianluca-Casagrande-Stiga** [comment]
Hi, I'm not able to reproduce the problem because I haven't found the reason yet. It happens 3-5 times a week with apparently no reason. I would like to know if anyone has encountered the same issue.
I'm working on it and as soon as I find a possible reason, I will update this issue.

**Comment by robertsLando** [comment]
What persistence/mqemitter are you using?

**Comment by casagrandeg** [comment]
```
import aedes from "aedes";
import tls from "tls";

const broker = aedes({
    authenticate: (client, username, password, callback) => {
    }
});

const port = 8883
const options = {
    key: process.env.MQTT_BROKER_CERT_KEY,
    cert: process.env.MQTT_BROKER_CERT_CRT,
    requestCert: true,
    ca: process.env.MQTT_CLIENT_CA
}

const server = tls.createServer(options, broker.handle);
server.listen(port, () => {
});

//broker events
broker.on('publish', async function (packet, client) {
});
broker.on('subscribe', async function (subscriptions, client) {
});
broker.on('unsubscribe', async function (unsubscriptions, client) {
});
broker.on('keepaliveTimeout', async function (client) {
});
broker.on('client', async function (client) {
});
broker.on('clientReady', async function (client) {
});
broker.on('clientDisconnect', async function (client) {
});
broker.on('connectionError', async function (client, error) {
});
broker.on('clientError', async function (client, error) {
});
broker.on('closed', async function () {
});
```

Hi Daniel, this is the schema of my mqtt broker made with aedes.

**Comment by robertsLando** [comment]
Ok so you are using the in memory mqemitter/persistence, not redis or mongo right? How does clients connect to broker? Do they have the clean flag set to true or false? What qos?

**Comment by casagrandeg** [comment]
Yes, I'm using only the server memory. The pub/sub qos is always 0. I haven't set the clean flag, so I think it is the "true" default value, but I will check every client.

**Comment by robertsLando** [comment]
I suggest to double check the clean flag and be sure it set to `true` otherwise that could be the problem, the memory becomes full after a week and the broker stops processing new messages

**Comment by robertsLando** [comment]
@Gianluca-Casagrande-Stiga check Linkedin when you have time ;)

**Comment by Gianluca-Casagrande-Stiga** [comment]
Hi, we still haven't found the reason of this bug. I've only noticed a considerable reduction of this bug when I set requestCert "false" on tls server options, but it is still present.
I have a question: May the reason be due to the node.js version? What is the recommended node version for Aedes 0.46.1? We are using 12.16-alpine Docker image.
Thanks

**Comment by robertsLando** [comment]
@Gianluca-Casagrande-Stiga Did you checked the clean flag on clients connect? Also my suggestion is to check proocess memory and cpu usage to see if there is a memory leak somewhere.

**Comment by Gianluca-Casagrande-Stiga** [comment]
Yes, I've checked the clean flag is set on each client.
The server memory is constant over the time (more or less 50MB), so it isn't correlated to the bug event.

**Comment by robertsLando** [comment]
@getlarge Any clue what could cause this?

**Comment by JaosnHsieh** [comment]
Related https://github.com/moscajs/aedes/issues/553

**Comment by robertsLando** [comment]
@JaosnHsieh Your fix doesn't resolve the main bug, why ignoring $SYS emits works?

**Comment by robertsLando** [comment]
Wondering if it could be something related to this: https://github.com/moscajs/aedes/blob/d8b976a7bdfce5a7b8495a9aa3d96a9dcd4353b0/lib/handlers/subscribe.js#L100

**Comment by robertsLando** [comment]
@JaosnHsieh parallel alway call released once finished, check line 22, released is a fastparallel option

**Comment by robertsLando** [comment]
Just checked again my edited PR with your code and seems it's not working, I think there is a bug elsewhere so...

**Comment by robertsLando** [comment]
After some digging by adding some logs this is the state of the queue when it stucks:

```js
starting parallel.. current: 1
starting parallel.. queue: Packet {
  cmd: 'publish',
  brokerId: '5bdea25f-a04d-4d5f-a19d-ddbb1d104410',
  brokerCounter: 32,
  topic: '$SYS/5bdea25f-a04d-4d5f-a19d-ddbb1d104410/new/clients',
  payload: <Buffer 38 33 35 33 39 33>,
  qos: 0,
  retain: false,
  dup: false
}
starting parallel.. callback: [Function: release]
```

I will try to check fastparallel source code to see if I can find out more. BTW seems that by removing the publish on `new/clients` the bug doens't happen

**Comment by robertsLando** [comment]
running an npm update should fix the issue now

**Comment by mcollina** [comment]
Amazing @robertsLando, you have been able to track down this issue - this is kind of a degree in async debugging!

**Comment by JaosnHsieh** [comment]
> @JaosnHsieh Your fix doesn't resolve the main bug, why ignoring $SYS emits works?

Thanks for you fix first !! I'm still curious why it's related to `$SYS` ... `/new/clients` topic.

I found this change `done() to setImmediate(done)` in aedes also fixed the issue.

```
diff --git a/node_modules/aedes/aedes.js b/node_modules/aedes/aedes.js
index 5f74455..9a7547f 100755
--- a/node_modules/aedes/aedes.js
+++ b/node_modules/aedes/aedes.js
@@ -151,7 +151,7 @@ function Aedes (opts) {
     if (that.clients[clientId] && serverId !== that.id) {
       that.clients[clientId].close(done)
     } else {
-      done()
+      setImmediate(done)
     }
   })
 
```

I have the same confusion as this [issue](https://github.com/mcollina/fastparallel/issues/13) on fastparallel.

I thought `fastparallel` is faster mostly because interval v8 cache mechanism as the [doc](https://github.com/mcollina/fastparallel#why-it-is-so-fast).

It seems like we have to put asynchronous action call when using `fastparallel`. 
Do you have any ideas why it has to be like that?



you can reproduce the fix by

```
$ git clone --branch why-setimmediate-needed-on-fastparallel git@github.com:JaosnHsieh/aedes-unable-to-publish.git
$ npm i
$ npm run server
$ npm run client
```
